R/assertions.R

```diff
@@ -772,46 +773,66 @@ assert_s3_class <- function(arg, cls,
 #'   assert_list_of(list, "numeric", named = TRUE)
 #' }
 #' try(example_fun2(list(1, 2, 3, d = 4)))
-assert_list_of <- function(arg, class, named = FALSE, optional = TRUE) {
-  assert_character_scalar(class)
+assert_list_of <- function(arg, cls,
+                           named = FALSE,
+                           optional = TRUE,
+                           arg_name = rlang::caller_arg(arg),
+                           message = NULL,
+                           class = "assert_list_of",
+                           call = parent.frame()) {
+  assert_character_scalar(cls)
   assert_logical_scalar(named)
   assert_logical_scalar(optional)
 
   if (is.null(arg) && optional) {
     return(invisible(arg))
   }
 
-  assert_s3_class(arg, "list")
-
-  is_class <- map_lgl(arg, inherits, class) | map_chr(arg, typeof) == class
-  if (!all(is_class)) {
-    info_msg <- paste(
-      sprintf("\u2716 Element %s is %s", which(!is_class), map_chr(arg[!is_class], what_is_it)),
-      collapse = "\n"
-    )
-    err_msg <- sprintf(
-      "Each element of `%s` must be an object of class/type '%s' but the following are not:\n%s",
-      arg_name(substitute(arg)),
-      class,
-      info_msg
+  # check the passed argument is a list
+  assert_s3_class(
+    arg,
+    cls = "list",
+    arg_name = arg_name,
+    message = message,
+    class = class,
+    call = call
+  )
+  # if list must be named, check this
+  if (named) {
+    assert_named(
+      arg,
+      arg_name = arg_name,
+      message = message,
+      class = class,
+      call = call
     )
-    abort(err_msg)
   }
 
-  if (named && length(arg) > 0) {
-    if (is.null(names(arg))) {
-      abort(paste0(
-        "All elements of ", arg_name(substitute(arg)), " must be named.\n",
-        "No element is named."
-      ))
-    }
-    unnamed <- which(names(arg) == "")
-    if (length(unnamed) > 0) {
-      abort(paste0(
-        "All elements of ", arg_name(substitute(arg)), " must be named.\n",
-        "The following elements are not named: ", enumerate(unnamed, quote_fun = NULL)
-      ))
+  # check each list element is the expected class/type
+  is_class <- map_lgl(arg, inherits, cls) | map_chr(arg, typeof) == cls
+  if (!all(is_class)) {
+    # construct supplementary message listing elements that are not correct type
+    if (!is.null(message)) {
+      info_msg <- glue::glue_collapse(
+        glue::glue(
```

Comment:
if we need glue can list these out in the [admiraldev](https://github.com/pharmaverse/admiraldev/blob/main/R/admiraldev-package.R)

Reply:
added the glue functions to the imports